driver-core/src/main/com/mongodb/internal/connection/LoggingCommandEventSender.java

```diff
@@ -91,38 +113,41 @@ public void sendStartedEvent() {
         commandDocument = null;
     }
 
-    private String getTruncatedJsonCommand() {
-        StringWriter writer = new StringWriter();
-
-        try (BsonReader bsonReader = commandDocument.asBsonReader()) {
-            JsonWriter jsonWriter = new JsonWriter(writer,
-                    JsonWriterSettings.builder().outputMode(JsonMode.RELAXED).maxLength(MAX_COMMAND_DOCUMENT_LENGTH_TO_LOG).build());
-
-            jsonWriter.pipe(bsonReader);
-
-            if (jsonWriter.isTruncated()) {
-                writer.append(" ...");
-            }
-
-            return writer.toString();
-        }
-    }
-
     @Override
     public void sendFailedEvent(final Throwable t) {
         Throwable commandEventException = t;
         if (t instanceof MongoCommandException && redactionRequired) {
-            commandEventException = new MongoCommandException(new BsonDocument(), description.getServerAddress());
+            MongoCommandException originalCommandException = (MongoCommandException) t;
+            commandEventException = new MongoCommandException(new BsonDocument(), originalCommandException.getServerAddress());
+            commandEventException.setStackTrace(t.getStackTrace());
         }
         long elapsedTimeNanos = System.nanoTime() - startTimeNanos;
 
         if (loggingRequired()) {
-            logger.debug(
-                    format("Execution of command with request id %d failed to complete successfully in %s ms on connection [%s] "
-                                    + "to server %s",
-                            message.getId(), getElapsedTimeFormattedInMilliseconds(elapsedTimeNanos), description.getConnectionId(),
-                            description.getServerAddress()),
-                    commandEventException);
+            if (description.getServiceId() == null) {
+                logger.debug("Command failed", getClusterId(), commandEventException,
+                        "Command \"%s\" failed in %.2f ms using a connection with driver-generated ID %d and server-generated ID "
+                                + "%d to %s:%s. The request ID is %d.",
+                        "commandName", commandName,
+                        "durationMS", elapsedTimeNanos / NANOS_PER_MILLI,
```

Comment:
I propose to use `NANOSECONDS.toMillis(elapsedTimeNanos)` instead of reinventing it.

Reply:
It's done this way to avoid rounding, since partial milliseconds can be significant in high speed networks. Note that `NANOS_PER_MILLI` is a double not an int.